File: test/capture_intent_contract_test.dart
import 'package:ar_flutter_plugin_2/models/capture_intent_contract.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Issue 99 capture intent contract', () {
    test('preflight failures allocate no ID and automatic requires selector',
        () {
      final model = _admissionModel();
      expect(
        model
            .preflight(
                _intent(CaptureLane.automatic, selector: false), _liability())
            .state,
        CaptureIntentState.automaticWaitingSelector,
      );
      expect(
        model
            .preflight(_intent(CaptureLane.manual), _liability())
            .accepted!
            .identity
            .attemptOrdinal,
        1,
      );
    });

    test('automatic preserves protected manual lane and both coexist', () {
      final model = _admissionModel();
      expect(
        model
            .preflight(
              _intent(CaptureLane.automatic),
              _liability(memory: 70 * 1024 * 1024),
            )
            .reason,
        'protected-manual-capacity',
      );
      expect(
        model
            .preflight(_intent(CaptureLane.automatic), _liability())
            .isAccepted,
        isTrue,
      );
      expect(
        model.preflight(_intent(CaptureLane.manual), _liability()).isAccepted,
        isTrue,
      );
      expect(model.admittedCount, 2);
    });

    test('accepted receipt precedes exactly one exposure and complete commit',
        () {
      final accepted = _admissionModel()
          .preflight(_intent(CaptureLane.manual), _liability())
          .accepted!;
      final machine = CaptureAttemptReferenceMachine(accepted);
      expect(machine.receipt.durable, isTrue);
      expect(machine.requestExposure('expose').disposition,
          CaptureTransitionDisposition.applied);
      expect(machine.requestExposure('expose').disposition,
          CaptureTransitionDisposition.exactReplay);
      expect(machine.exposureCount, 1);
      final components = [
        _component(CaptureComponentKind.jpeg),
        _component(CaptureComponentKind.dng)
      ];
      machine.ownSensorOutput('output', components);
      expect(machine.validate('validate').disposition,
          CaptureTransitionDisposition.applied);
      expect(
          machine
              .prepareDurable('prepare', _commit(accepted, components))
              .disposition,
          CaptureTransitionDisposition.applied);
      expect(machine.retainedImageBytes, 0);
      expect(
        machine
            .commit('commit',
                captureId: 'capture',
                captureRevision: 1,
                manifestId: 'manifest')
            .receipt
            .terminal!
            .kind,
        CaptureTerminalKind.committedPicture,
      );
      expect(
        machine
            .commit('commit',
                captureId: 'capture',
                captureRevision: 1,
                manifestId: 'manifest')
            .disposition,
        CaptureTransitionDisposition.exactReplay,
      );
      expect(
        machine
            .commit('changed',
                captureId: 'other', captureRevision: 2, manifestId: 'other')
            .disposition,
        CaptureTransitionDisposition.conflict,
      );
    });

    test('incomplete components reject before commit', () {
      final accepted = _admissionModel()
          .preflight(_intent(CaptureLane.manual), _liability())
          .accepted!;
      final machine = CaptureAttemptReferenceMachine(accepted)
        ..requestExposure('expose');
      machine
          .ownSensorOutput('output', [_component(CaptureComponentKind.jpeg)]);
      expect(machine.validate('validate').reason, 'incomplete-component-set');
      expect(machine.receipt.phase, CaptureAttemptPhase.sensorOutputOwned);
    });

    test(
        'timeout remains unknown, abandonment fences late work, later succeeds',
        () {
      final model = _admissionModel();
      final first =
          model.preflight(_intent(CaptureLane.manual), _liability()).accepted!;
      final failed = CaptureAttemptReferenceMachine(first)
        ..requestExposure('expose');
      expect(failed.timeoutUnknown('timeout').receipt.phase,
          CaptureAttemptPhase.exposureRequested);
      expect(failed.outcomeUnknown, isTrue);
      expect(
          failed
              .abandon('abandon', 'proven-absent')
              .receipt
              .terminal!
              .metadataOnly,
          isTrue);
      expect(failed.lateCallback('late').reason, 'late-producer-fenced');
      model.release(first.identity);
      final later =
          model.preflight(_intent(CaptureLane.manual), _liability()).accepted!;
      expect(later.identity.attemptOrdinal, 2);
    });

    test('reservation and ordinal overflow abandon or reject without wrap', () {
      final model = _admissionModel();
      final accepted = model
          .preflight(_intent(CaptureLane.manual), _liability(store: 100))
          .accepted!;
      final machine = CaptureAttemptReferenceMachine(accepted)
        ..requestExposure('expose');
      final components = [
        _component(CaptureComponentKind.jpeg, bytes: 80),
        _component(CaptureComponentKind.dng, bytes: 80),
      ];
      machine.ownSensorOutput('output', components);
      machine.validate('validate');
      expect(
        machine
            .prepareDurable('prepare', _commit(accepted, components))
            .receipt
            .terminal!
            .reason,
        'reservation-overflow',
      );
      final exhausted = _admissionModel()
        ..setNextOrdinalForTest(capturePortableOrdinalMaximum + 1);
      expect(
        exhausted.preflight(_intent(CaptureLane.manual), _liability()).reason,
        'ordinal-or-liability-overflow',
      );
    });

    test('generated fault lifecycle matrix is the complete cross-product', () {
      final matrix = CaptureFaultLifecycleMatrix.generate();
      expect(
          matrix,
          hasLength(CaptureLane.values.length *
              5 *
              CaptureFault.values.length *
              CaptureLifecycleEvent.values.length));
      expect(matrix.map((value) => value.phase).toSet(),
          CaptureAttemptPhase.values.take(5).toSet());
      expect(matrix.map((value) => value.lane).toSet(),
          CaptureLane.values.toSet());
      expect(matrix.map((value) => value.fault).toSet(),
          CaptureFault.values.toSet());
      expect(matrix.map((value) => value.lifecycleEvent).toSet(),
          CaptureLifecycleEvent.values.toSet());
      expect(matrix.map((value) => value.expectedOutcome).toSet(),
          CaptureMatrixOutcome.values.toSet());
    });

    test('transition table rejects every edge outside the canonical graph', () {
      final pairs = [
        for (final from in CaptureAttemptPhase.values)
          for (final to in CaptureAttemptPhase.values) (from, to),
      ];
      expect(
        pairs
            .where((pair) =>
                CaptureAttemptTransitionTable.isLegal(pair.$1, pair.$2))
            .length,
        10,
      );
      expect(
        pairs
            .where((pair) =>
                !CaptureAttemptTransitionTable.isLegal(pair.$1, pair.$2))
            .length,
        39,
      );
    });

    test('overflow releases the lane and a later full commit succeeds', () {
      final model = _admissionModel();
      final failed = model
          .preflight(_intent(CaptureLane.manual), _liability(store: 100))
          .accepted!;
      final failedMachine = CaptureAttemptReferenceMachine(failed)
        ..requestExposure('expose');
      final oversized = [
        _component(CaptureComponentKind.jpeg, bytes: 80),
        _component(CaptureComponentKind.dng, bytes: 80),
      ];
      failedMachine
        ..ownSensorOutput('output', oversized)
        ..validate('validate')
        ..prepareDurable('prepare', _commit(failed, oversized));
      model.release(failed.identity);
      final later =
          model.preflight(_intent(CaptureLane.manual), _liability()).accepted!;
      final laterMachine = CaptureAttemptReferenceMachine(later)
        ..requestExposure('later-expose');
      final complete = [
        _component(CaptureComponentKind.jpeg),
        _component(CaptureComponentKind.dng),
      ];
      laterMachine
        ..ownSensorOutput('later-output', complete)
        ..validate('later-validate')
        ..prepareDurable('later-prepare', _commit(later, complete));
      expect(
        laterMachine
            .commit(
              'later-commit',
              captureId: 'later-capture',
              captureRevision: 1,
              manifestId: 'later-manifest',
            )
            .receipt
            .terminal!
            .kind,
        CaptureTerminalKind.committedPicture,
      );
    });

    test('DTO value equality hashes deep lists and copies mutable bytes', () {
      final bytes = _digest('mutable');
      final first = CaptureComponentDescriptor(
          kind: CaptureComponentKind.jpeg,
          byteLength: 40,
          sha256: bytes,
          durableObjectId: 'object');
      final second = CaptureComponentDescriptor(
          kind: CaptureComponentKind.jpeg,
          byteLength: 40,
          sha256: List<int>.of(bytes),
          durableObjectId: 'object');
      expect(first, second);
      expect(first.hashCode, second.hashCode);
      bytes[0] ^= 0xff;
      expect(first, second);
      expect(
          _commit(_accepted(CaptureLane.manual), [
            _component(CaptureComponentKind.jpeg),
            _component(CaptureComponentKind.dng)
          ]),
          _commit(_accepted(CaptureLane.manual), [
            _component(CaptureComponentKind.jpeg),
            _component(CaptureComponentKind.dng)
          ]));
    });

    test('numeric domains reject negative overflow and malformed digests', () {
      expect(
          () => CaptureComponentDescriptor(
              kind: CaptureComponentKind.jpeg,
              byteLength: -1,
              sha256: _digest('x'),
              durableObjectId: 'x'),
          throwsArgumentError);
      expect(
          () => CaptureReservationLiability(
              memoryBytes: -1,
              physicalStoreBytes: 0,
              componentEntries: 0,
              terminalEntries: 0,
              rollbackBytes: 0,
              physicallyBacked: true),
          throwsArgumentError);
      expect(
          () => CaptureReservationLiability(
              memoryBytes: 0,
              physicalStoreBytes: 0,
              componentEntries: capturePortableEntryMaximum + 1,
              terminalEntries: 0,
              rollbackBytes: 0,
              physicallyBacked: true),
          throwsArgumentError);
      expect(
          () => CaptureAttemptIdentity(
              attemptId: 'a',
              commitId: 'c',
              attemptOrdinal: 0,
              lifecycleCut: _cut()),
          throwsArgumentError);
      expect(
          () => CaptureLifecycleCut(
              sessionId: 's',
              sessionGeneration: -1,
              groupId: 'g',
              groupGeneration: 0,
              arSessionId: 'a',
              viewId: 'v',
              viewGeneration: 0,
              bindingToken: 'b',
              lifecycleSequence: 0,
              operationGeneration: 0),
          throwsArgumentError);
      expect(
          () => CaptureComponentDescriptor(
              kind: CaptureComponentKind.jpeg,
              byteLength: 1,
              sha256: [1],
              durableObjectId: 'x'),
          throwsArgumentError);
    });

    test('finalizer scheduler gives manual running priority and promotes once',
        () {
      final manual = _accepted(CaptureLane.manual);
      final automatic = _accepted(CaptureLane.automatic, ordinal: 2);
      final scheduler = CaptureFinalizerScheduler();
      final assignments =
          scheduler.scheduleReady([automatic, manual, automatic]);
      expect(assignments.map((value) => value.position), [
        CaptureFinalizerPosition.running,
        CaptureFinalizerPosition.fundedWaiting,
        CaptureFinalizerPosition.rejected
      ]);
      expect(scheduler.running, manual);
      expect(scheduler.fundedWaiting, automatic);
      expect(scheduler.ownsExposure(automatic.identity), isFalse);
      expect(CaptureAttemptReferenceMachine(automatic).exposureCount, 0);
      expect(scheduler.release(manual.identity), automatic);
      expect(scheduler.ownsExposure(automatic.identity), isTrue);
      expect(scheduler.release(automatic.identity), isNull);
    });

    test('all 540 rows execute canonical terminal query and replay outcomes',
        () {
      final outcomeCounts = <CaptureMatrixOutcome, int>{};
      final accepted = {
        CaptureLane.manual: _accepted(CaptureLane.manual),
        CaptureLane.automatic: _accepted(CaptureLane.automatic, ordinal: 2),
      };
      for (final row in CaptureFaultLifecycleMatrix.generate()) {
        final attempt = accepted[row.lane]!;
        final request = _commit(attempt, [
          _component(CaptureComponentKind.jpeg),
          _component(CaptureComponentKind.dng)
        ]);
        final execution =
            CaptureFaultLifecycleMatrix.execute(row, attempt, request);
        expect(execution.outcome, row.expectedOutcome,
            reason:
                '${row.lane}/${row.phase}/${row.fault}/${row.lifecycleEvent}');
        expect(execution.exactReplay, isTrue);
        expect(execution.changedReplayConflict, isTrue);
        expect(execution.lifecycleFenceNoOp, isTrue);
        expect(
            execution.lateCallbackNoOp, row.fault == CaptureFault.lateCallback);
        expect(execution.exposureCount,
            row.phase == CaptureAttemptPhase.reservedAccepted ? 0 : 1);
        outcomeCounts.update(execution.outcome, (value) => value + 1,
            ifAbsent: () => 1);
      }
      expect(outcomeCounts, {
        CaptureMatrixOutcome.abandoned: 396,
        CaptureMatrixOutcome.outcomeUnknown: 108,
        CaptureMatrixOutcome.committed: 36,
      });

      final committed = CaptureFaultLifecycleCase(
          CaptureLane.manual,
          CaptureAttemptPhase.durablePrepared,
          CaptureFault.timeout,
          CaptureLifecycleEvent.backgrounded,
          CaptureMatrixOutcome.committed);
      final mutated = CaptureFaultLifecycleCase(
          committed.lane,
          committed.phase,
          CaptureFault.componentFailure,
          committed.lifecycleEvent,
          CaptureMatrixOutcome.abandoned);
      final attempt = accepted[CaptureLane.manual]!;
      expect(
          CaptureFaultLifecycleMatrix.execute(
              committed,
              attempt,
              _commit(attempt, [
                _component(CaptureComponentKind.jpeg),
                _component(CaptureComponentKind.dng)
              ])).outcome,
          isNot(CaptureFaultLifecycleMatrix.execute(
              mutated,
              attempt,
              _commit(attempt, [
                _component(CaptureComponentKind.jpeg),
                _component(CaptureComponentKind.dng)
              ])).outcome));
    });
  });
}

CaptureIntent _intent(CaptureLane lane, {bool selector = true}) =>
    CaptureIntent(
      lane: lane,
      lifecycleCut: CaptureLifecycleCut(
        sessionId: 'session',
        sessionGeneration: 1,
        groupId: 'group',
        groupGeneration: 1,
        arSessionId: 'ar',
        viewId: 'view',
        viewGeneration: 1,
        bindingToken: 'binding',
        lifecycleSequence: 1,
        operationGeneration: 1,
      ),
      profile: CaptureComponentProfile(
        profileId: 'raw+jpeg',
        requiredComponents: const {
          CaptureComponentKind.jpeg,
          CaptureComponentKind.dng
        },
        maximumComponentBytes: 200,
        maximumWorkingBytes: 60 * 1024 * 1024,
      ),
      selectorValid: selector,
      trackingValid: true,
      captureHealthy: true,
      durabilityPreflightValid: true,
      canonicalIntentHash: _digest('intent-${lane.name}'),
    );

CaptureAdmissionReferenceModel _admissionModel() =>
    CaptureAdmissionReferenceModel(
      protectedManualMemoryBytes: 64 * 1024 * 1024,
    );

CaptureLifecycleCut _cut() => CaptureLifecycleCut(
      sessionId: 'session',
      sessionGeneration: 1,
      groupId: 'group',
      groupGeneration: 1,
      arSessionId: 'ar',
      viewId: 'view',
      viewGeneration: 1,
      bindingToken: 'binding',
      lifecycleSequence: 1,
      operationGeneration: 1,
    );

CaptureAcceptedAttempt _accepted(CaptureLane lane, {int ordinal = 1}) =>
    CaptureAcceptedAttempt(
      identity: CaptureAttemptIdentity(
          attemptId: 'attempt-$ordinal',
          commitId: 'commit-$ordinal',
          attemptOrdinal: ordinal,
          lifecycleCut: _cut()),
      lane: lane,
      profile: CaptureComponentProfile(
          profileId: 'raw+jpeg',
          requiredComponents: const {
            CaptureComponentKind.jpeg,
            CaptureComponentKind.dng
          },
          maximumComponentBytes: 200,
          maximumWorkingBytes: 1024),
      reservation: CaptureReservationLiability(
          memoryBytes: 1024,
          physicalStoreBytes: 200,
          componentEntries: 2,
          terminalEntries: 1,
          rollbackBytes: 0,
          physicallyBacked: true),
      canonicalIntentHash: _digest('intent-$ordinal'),
      acceptedReceiptHash: _digest('accepted-$ordinal'),
    );

CaptureReservationLiability _liability(
        {int memory = 60 * 1024 * 1024, int store = 200}) =>
    CaptureReservationLiability(
      memoryBytes: memory,
      physicalStoreBytes: store,
      componentEntries: 2,
      terminalEntries: 1,
      rollbackBytes: 0,
      physicallyBacked: true,
    );

CaptureComponentDescriptor _component(CaptureComponentKind kind,
        {int bytes = 40}) =>
    CaptureComponentDescriptor(
      kind: kind,
      byteLength: bytes,
      sha256: _digest('sha-${kind.name}-$bytes'),
      durableObjectId: 'object-${kind.name}',
    );

CaptureCommitRequest _commit(CaptureAcceptedAttempt accepted,
        List<CaptureComponentDescriptor> components) =>
    CaptureCommitRequest(
      accepted: accepted,
      components: components,
      exposureTimestampNanoseconds: 10,
      poseRecordHash: _digest('pose'),
      cameraModelHash: _digest('camera'),
      validationRecordHash: _digest('validation'),
      ledgerRecordHash: _digest('ledger'),
    );

List<int> _digest(String value) => List<int>.generate(
      32,
      (index) => value.codeUnitAt(index % value.length) ^ index,
      growable: false,
    );

File: android/src/test/kotlin/com/uhg0/ar_flutter_plugin_2/capture/CaptureIntentContractTest.kt
package com.uhg0.ar_flutter_plugin_2.capture

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class CaptureIntentContractTest {
    @Test
    fun `native reservation bounds include durable metadata coexistence and rollback with checked arithmetic`() {
        val physical = NativeCaptureReservationBoundsV2.physicalBytes(1024, 2)
        assertTrue(physical > 1024 + NativeCaptureReservationBoundsV2.COEXISTENCE_BYTES)
        assertEquals(1024 + NativeCaptureReservationBoundsV2.ROLLBACK_METADATA_BYTES,
            NativeCaptureReservationBoundsV2.rollbackBytes(1024))
        assertFails { NativeCaptureReservationBoundsV2.physicalBytes(Long.MAX_VALUE, 2) }
        assertFails { NativeCaptureReservationBoundsV2.rollbackBytes(Long.MAX_VALUE) }
    }
    @Test
    fun `accepted record is durable before exactly one exposure and one committed terminal`() {
        val accepted = accepted()
        val machine = CaptureAttemptReferenceMachine(accepted)
        assertTrue(machine.receipt.durable)
        assertEquals(CaptureTransitionDisposition.APPLIED, machine.requestExposure("expose").disposition)
        assertEquals(CaptureTransitionDisposition.EXACT_REPLAY, machine.requestExposure("expose").disposition)
        assertEquals(1, machine.exposureCount)
        val components = components()
        machine.ownSensorOutput("output", components)
        machine.validate("validate")
        assertEquals(CaptureTransitionDisposition.APPLIED, machine.prepareDurable("prepare", commit(accepted, components)).disposition)
        assertEquals(0, machine.retainedImageBytes)
        val terminal = machine.commit("commit", "capture", 1, "manifest")
        assertEquals(CaptureTerminalKind.COMMITTED_PICTURE, terminal.receipt.terminal?.kind)
        assertEquals(CaptureTransitionDisposition.EXACT_REPLAY, machine.commit("commit", "capture", 1, "manifest").disposition)
        assertEquals(CaptureTransitionDisposition.CONFLICT, machine.commit("changed", "other", 2, "other").disposition)
    }

    @Test
    fun `timeout is unknown and overflow abandons metadata only`() {
        val accepted = accepted(storeBytes = 100)
        val machine = CaptureAttemptReferenceMachine(accepted)
        machine.requestExposure("expose")
        assertEquals(CaptureAttemptPhase.EXPOSURE_REQUESTED, machine.timeoutUnknown().receipt.phase)
        assertTrue(machine.outcomeUnknown)
        val components = components(80)
        machine.ownSensorOutput("output", components)
        machine.validate("validate")
        val terminal = machine.prepareDurable("prepare", commit(accepted, components)).receipt.terminal
        assertEquals(CaptureTerminalKind.ABANDONED_ATTEMPT, terminal?.kind)
        assertEquals("reservation-overflow", terminal?.reason)
        assertEquals(0, machine.retainedImageBytes)
        assertFalse(machine.outcomeUnknown)
        assertEquals("late-producer-fenced", machine.lateCallback().reason)
    }

    @Test
    fun `generated matrix covers every preterminal fault and lifecycle cut`() {
        val matrix = CaptureFaultLifecycleMatrix.generate()
        assertEquals(CaptureLane.entries.size * 5 * CaptureFault.entries.size * CaptureLifecycleEvent.entries.size, matrix.size)
        assertEquals(CaptureLane.entries.toSet(), matrix.map { it.lane }.toSet())
        assertEquals(CaptureAttemptPhase.entries.take(5).toSet(), matrix.map { it.phase }.toSet())
        assertEquals(CaptureFault.entries.toSet(), matrix.map { it.fault }.toSet())
        assertEquals(CaptureLifecycleEvent.entries.toSet(), matrix.map { it.lifecycleEvent }.toSet())
        assertEquals(CaptureMatrixOutcome.entries.toSet(), matrix.map { it.expectedOutcome }.toSet())
    }

    @Test
    fun `transition table rejects every noncanonical edge`() {
        val pairs = CaptureAttemptPhase.entries.flatMap { from ->
            CaptureAttemptPhase.entries.map { destination -> Pair(from, destination) }
        }
        assertEquals(10, pairs.count { CaptureAttemptTransitionTable.isLegal(it.first, it.second) })
        assertEquals(39, pairs.count { !CaptureAttemptTransitionTable.isLegal(it.first, it.second) })
    }

    @Test
    fun `intent numeric domains and digests reject malformed values`() {
        val intent = CaptureIntent(CaptureLane.AUTOMATIC, lifecycle(), profile(), true, true, true, true, digest("intent"))
        assertEquals(CaptureLane.AUTOMATIC, intent.lane)
        assertEquals(CaptureIntentState.entries.size, 8)
        assertFails { CaptureComponentDescriptor(CaptureComponentKind.JPEG, -1, digest("jpeg"), "object") }
        assertFails { CaptureReservationLiability(-1, 0, 0, 0, 0, true) }
        assertFails { CaptureReservationLiability(0, 0, CAPTURE_PORTABLE_ENTRY_MAXIMUM + 1, 0, 0, true) }
        assertFails { CaptureAttemptIdentity("a", "c", 0, lifecycle()) }
        assertFails { CaptureLifecycleCut("s", -1, "g", 0, "a", "v", 0, "b", 0, 0) }
        assertFails { CaptureComponentDescriptor(CaptureComponentKind.JPEG, 1, listOf(1), "object") }
        assertFails { CaptureReservationLiability(0, Long.MAX_VALUE, 0, 0, 1, true) }
    }

    @Test
    fun `contract values defensively copy collections and retain structural equality`() {
        val mutableDigest = digest("intent").toMutableList()
        val firstIntent = CaptureIntent(CaptureLane.AUTOMATIC, lifecycle(), profile(), true, true, true, true, mutableDigest)
        val equalIntent = CaptureIntent(CaptureLane.AUTOMATIC, lifecycle(), profile(), true, true, true, true, mutableDigest.toList())
        val stableHash = firstIntent.hashCode()
        val machine = CaptureAttemptReferenceMachine(CaptureAcceptedAttempt(
            CaptureAttemptIdentity("immutable", "immutable-commit", 7, lifecycle()), CaptureLane.AUTOMATIC,
            profile(), CaptureReservationLiability(10, 10, 2, 1, 0, true), mutableDigest, digest("receipt")))
        val stableReceipt = machine.receipt.receiptHash
        mutableDigest[0] = mutableDigest[0] xor 0xff
        assertEquals(equalIntent, firstIntent)
        assertEquals(stableHash, firstIntent.hashCode())
        assertEquals(stableReceipt, machine.receipt.receiptHash)
        assertFails { (firstIntent.canonicalIntentHash as MutableList<Int>)[0] = 0 }

        val mutableComponents = components().toMutableList()
        val accepted = accepted()
        val request = CaptureCommitRequest(accepted, mutableComponents, 10, digest("pose"), digest("camera"),
            digest("validation"), digest("ledger"))
        val equalRequest = commit(accepted, components())
        val requestHash = request.hashCode()
        mutableComponents.clear()
        assertEquals(2, request.components.size)
        assertEquals(equalRequest, request)
        assertEquals(requestHash, request.hashCode())
        assertFails { (request.components as MutableList<CaptureComponentDescriptor>).clear() }
        assertEquals(profile(), profile())
        assertEquals(profile().hashCode(), profile().hashCode())
    }

    @Test
    fun `scheduler gives manual running priority and promotes deterministically`() {
        val manual = accepted(CaptureLane.MANUAL, 1)
        val automatic = accepted(CaptureLane.AUTOMATIC, 2)
        val scheduler = CaptureFinalizerScheduler()
        val assignments = scheduler.scheduleReady(listOf(automatic, manual, automatic))
        assertEquals(listOf(CaptureFinalizerPosition.RUNNING, CaptureFinalizerPosition.FUNDED_WAITING, CaptureFinalizerPosition.REJECTED), assignments.map { it.position })
        assertEquals(manual, scheduler.running)
        assertEquals(automatic, scheduler.fundedWaiting)
        assertFalse(scheduler.ownsExposure(automatic.identity))
        assertEquals(0, CaptureAttemptReferenceMachine(automatic).exposureCount)
        assertEquals(automatic, scheduler.release(manual.identity))
        assertTrue(scheduler.ownsExposure(automatic.identity))
        assertEquals(null, scheduler.release(automatic.identity))
    }

    @Test
    fun `all 540 rows execute canonical query terminal and replay outcomes`() {
        val attempts = mapOf(CaptureLane.MANUAL to accepted(CaptureLane.MANUAL, 1), CaptureLane.AUTOMATIC to accepted(CaptureLane.AUTOMATIC, 2))
        val counts = mutableMapOf<CaptureMatrixOutcome, Int>()
        val lifecycleEffects = mutableMapOf<CaptureLifecycleEvent, MutableSet<CaptureLifecycleOwnershipEffect>>()
        CaptureFaultLifecycleMatrix.generate().forEach { row ->
            val attempt = requireNotNull(attempts[row.lane])
            val execution = CaptureFaultLifecycleMatrix.execute(row, attempt, commit(attempt, components()))
            assertEquals(row.expectedOutcome, execution.outcome)
            assertTrue(execution.exactReplay)
            assertTrue(execution.changedReplayConflict)
            lifecycleEffects.getOrPut(row.lifecycleEvent) { mutableSetOf() }.add(execution.ownershipEffect)
            assertEquals(row.fault == CaptureFault.LATE_CALLBACK, execution.lateCallbackNoOp)
            assertEquals(if (row.phase == CaptureAttemptPhase.RESERVED_ACCEPTED &&
                execution.outcome != CaptureMatrixOutcome.COMMITTED) 0 else 1, execution.exposureCount)
            assertEquals(row.phase.ordinal >= CaptureAttemptPhase.SENSOR_OUTPUT_OWNED.ordinal,
                execution.sensorOutputOwnedAtCut)
            val activeOwnerContinues = row.lifecycleEvent == CaptureLifecycleEvent.AUTOMATIC_DISABLED &&
                row.fault == CaptureFault.LATE_CALLBACK
            assertEquals(activeOwnerContinues && row.phase == CaptureAttemptPhase.RESERVED_ACCEPTED,
                execution.continuedExposureAfterCut)
            assertEquals(activeOwnerContinues && row.phase.ordinal < CaptureAttemptPhase.SENSOR_OUTPUT_OWNED.ordinal,
                execution.sensorOutputAcquiredAfterCut)
            assertTrue(execution.classificationReason.contains(execution.ownershipEffect.name))
            counts[execution.outcome] = (counts[execution.outcome] ?: 0) + 1
        }
        assertEquals(mapOf(CaptureMatrixOutcome.ABANDONED to 378, CaptureMatrixOutcome.OUTCOME_UNKNOWN to 90,
            CaptureMatrixOutcome.COMMITTED to 72), counts)
        assertEquals(mapOf(
            CaptureLifecycleEvent.AUTOMATIC_DISABLED to setOf(CaptureLifecycleOwnershipEffect.AUTOMATIC_ACTIVE_OWNER_CONTINUES),
            CaptureLifecycleEvent.ROUTE_LEFT to setOf(CaptureLifecycleOwnershipEffect.ROUTE_PRE_OUTPUT_CLASSIFIED,
                CaptureLifecycleOwnershipEffect.ROUTE_OWNED_OUTPUT_TRANSFERRED, CaptureLifecycleOwnershipEffect.ROUTE_DURABLE_STORE_OWNED),
            CaptureLifecycleEvent.BACKGROUNDED to setOf(CaptureLifecycleOwnershipEffect.BACKGROUND_PRE_OUTPUT_CLASSIFIED,
                CaptureLifecycleOwnershipEffect.BACKGROUND_OWNED_OUTPUT_FINISHING, CaptureLifecycleOwnershipEffect.BACKGROUND_DURABLE_STORE_OWNED),
            CaptureLifecycleEvent.VIEW_REPLACED to setOf(CaptureLifecycleOwnershipEffect.VIEW_PRE_OUTPUT_FENCED,
                CaptureLifecycleOwnershipEffect.VIEW_OWNED_OUTPUT_TRANSFERRED, CaptureLifecycleOwnershipEffect.VIEW_DURABLE_STORE_OWNED),
            CaptureLifecycleEvent.AR_SESSION_REPLACED to setOf(CaptureLifecycleOwnershipEffect.AR_PRE_OUTPUT_GROUP_FENCED,
                CaptureLifecycleOwnershipEffect.AR_OWNED_OUTPUT_FROZEN_GROUP, CaptureLifecycleOwnershipEffect.AR_DURABLE_OLD_GROUP_STORE_OWNED),
            CaptureLifecycleEvent.PROCESS_RESTARTED to setOf(CaptureLifecycleOwnershipEffect.PROCESS_PRE_DURABLE_ABSENT,
                CaptureLifecycleOwnershipEffect.PROCESS_DURABLE_RECEIPT_RECOVERED),
        ), lifecycleEffects)

        val attempt = requireNotNull(attempts[CaptureLane.MANUAL])
        val committed = CaptureFaultLifecycleCase(CaptureLane.MANUAL, CaptureAttemptPhase.DURABLE_PREPARED,
            CaptureFault.TIMEOUT, CaptureLifecycleEvent.BACKGROUNDED, CaptureMatrixOutcome.COMMITTED)
        val mutated = committed.copy(fault = CaptureFault.COMPONENT_FAILURE, expectedOutcome = CaptureMatrixOutcome.ABANDONED)
        assertFalse(CaptureFaultLifecycleMatrix.execute(committed, attempt, commit(attempt, components())).outcome ==
            CaptureFaultLifecycleMatrix.execute(mutated, attempt, commit(attempt, components())).outcome)

        val background = CaptureFaultLifecycleCase(CaptureLane.MANUAL, CaptureAttemptPhase.SENSOR_OUTPUT_OWNED,
            CaptureFault.LATE_CALLBACK, CaptureLifecycleEvent.BACKGROUNDED, CaptureMatrixOutcome.COMMITTED)
        val restarted = background.copy(lifecycleEvent = CaptureLifecycleEvent.PROCESS_RESTARTED,
            expectedOutcome = CaptureMatrixOutcome.ABANDONED)
        val backgroundExecution = CaptureFaultLifecycleMatrix.execute(background, attempt, commit(attempt, components()))
        val restartExecution = CaptureFaultLifecycleMatrix.execute(restarted, attempt, commit(attempt, components()))
        assertEquals(CaptureMatrixOutcome.COMMITTED, backgroundExecution.outcome)
        assertEquals(CaptureMatrixOutcome.ABANDONED, restartExecution.outcome)
        assertFalse(backgroundExecution.ownershipEffect == restartExecution.ownershipEffect)

        fun classify(event: CaptureLifecycleEvent, phase: CaptureAttemptPhase): CaptureMatrixExecution =
            CaptureFaultLifecycleMatrix.execute(CaptureFaultLifecycleCase(CaptureLane.MANUAL, phase,
                CaptureFault.LATE_CALLBACK, event, CaptureMatrixOutcome.ABANDONED), attempt, commit(attempt, components()))

        val automaticPre = classify(CaptureLifecycleEvent.AUTOMATIC_DISABLED, CaptureAttemptPhase.RESERVED_ACCEPTED)
        val routePre = classify(CaptureLifecycleEvent.ROUTE_LEFT, CaptureAttemptPhase.RESERVED_ACCEPTED)
        assertEquals(CaptureMatrixOutcome.COMMITTED, automaticPre.outcome)
        assertEquals(1, automaticPre.exposureCount)
        assertTrue(automaticPre.continuedExposureAfterCut)
        assertTrue(automaticPre.sensorOutputAcquiredAfterCut)
        assertEquals(CaptureMatrixOutcome.ABANDONED, routePre.outcome)
        assertEquals(0, routePre.exposureCount)
        assertFalse(routePre.sensorOutputAcquiredAfterCut)

        listOf(CaptureLifecycleEvent.ROUTE_LEFT, CaptureLifecycleEvent.BACKGROUNDED,
            CaptureLifecycleEvent.VIEW_REPLACED, CaptureLifecycleEvent.AR_SESSION_REPLACED).forEach { event ->
            val requested = classify(event, CaptureAttemptPhase.EXPOSURE_REQUESTED)
            assertEquals(CaptureMatrixOutcome.ABANDONED, requested.outcome)
            assertEquals(1, requested.exposureCount)
            assertFalse(requested.sensorOutputOwnedAtCut)
            assertFalse(requested.sensorOutputAcquiredAfterCut)
            assertTrue(requested.exactReplay)
            assertTrue(requested.changedReplayConflict)
        }

        val backgroundTimeout = CaptureFaultLifecycleMatrix.execute(CaptureFaultLifecycleCase(CaptureLane.MANUAL,
            CaptureAttemptPhase.EXPOSURE_REQUESTED, CaptureFault.TIMEOUT, CaptureLifecycleEvent.BACKGROUNDED,
            CaptureMatrixOutcome.OUTCOME_UNKNOWN), attempt, commit(attempt, components()))
        assertEquals(CaptureMatrixOutcome.OUTCOME_UNKNOWN, backgroundTimeout.outcome)
        assertEquals(CaptureAttemptPhase.EXPOSURE_REQUESTED, backgroundTimeout.phase)
        assertTrue(backgroundTimeout.classificationReason.contains("query-same-identity"))
        assertFalse(backgroundTimeout.sensorOutputAcquiredAfterCut)

        val viewOwned = classify(CaptureLifecycleEvent.VIEW_REPLACED, CaptureAttemptPhase.SENSOR_OUTPUT_OWNED)
        val arOwned = classify(CaptureLifecycleEvent.AR_SESSION_REPLACED, CaptureAttemptPhase.SENSOR_OUTPUT_OWNED)
        assertEquals(CaptureLifecycleOwnershipEffect.VIEW_OWNED_OUTPUT_TRANSFERRED, viewOwned.ownershipEffect)
        assertEquals(CaptureLifecycleOwnershipEffect.AR_OWNED_OUTPUT_FROZEN_GROUP, arOwned.ownershipEffect)
        assertFalse(viewOwned.classificationReason == arOwned.classificationReason)
        CaptureLifecycleEvent.entries.forEach { event ->
            val durable = classify(event, CaptureAttemptPhase.DURABLE_PREPARED)
            assertEquals(CaptureMatrixOutcome.COMMITTED, durable.outcome)
            assertTrue(durable.sensorOutputOwnedAtCut)
            assertFalse(durable.sensorOutputAcquiredAfterCut)
        }
    }

    private fun accepted(storeBytes: Long = 200) = accepted(CaptureLane.MANUAL, 1, storeBytes)

    private fun accepted(lane: CaptureLane, ordinal: Long, storeBytes: Long = 200) = CaptureAcceptedAttempt(
        identity = CaptureAttemptIdentity("attempt-$ordinal", "commit-$ordinal", ordinal, lifecycle()),
        lane = lane,
        profile = profile(),
        reservation = CaptureReservationLiability(1024, storeBytes, 2, 1, 0, true),
        canonicalIntentHash = digest("intent"),
        acceptedReceiptHash = digest("accepted"),
    )

    private fun lifecycle() = CaptureLifecycleCut("session", 1, "group", 1, "ar", "view", 1, "binding", 1, 1)

    private fun profile() = CaptureComponentProfile("raw+jpeg", setOf(CaptureComponentKind.JPEG, CaptureComponentKind.DNG), 200, 1024)

    private fun components(bytes: Long = 40) = listOf(
        CaptureComponentDescriptor(CaptureComponentKind.JPEG, bytes, digest("jpeg"), "jpeg-object"),
        CaptureComponentDescriptor(CaptureComponentKind.DNG, bytes, digest("dng"), "dng-object"),
    )

    private fun commit(accepted: CaptureAcceptedAttempt, components: List<CaptureComponentDescriptor>) =
        CaptureCommitRequest(accepted, components, 10, digest("pose"), digest("camera"), digest("validation"), digest("ledger"))

    private fun digest(value: String) = List(32) { index -> value[index % value.length].code xor index }

    private fun assertFails(block: () -> Unit) {
        try { block(); throw AssertionError("expected failure") } catch (_: RuntimeException) { }
    }
}
